Hash Cache Validation

danieldresser-ie opened the pull request:
Some experiments in the direction John was suggesting.  Is this anything like what you were thinking?

The way I've split the dirty range in half doesn't feel the most intuitive, but I'm not sure what would be clearer.

This does appear to work ( it catches all the same bugs in old IERendering as the old code ), except that it reports the failures in the wrong place ( the newly added test demonstrates this by failing ).

The failures being reported in the wrong place is due to a general problem with error reporting in the hash cache.

This can be demonstrated in current Gaffer with a scene like this:

```
import Gaffer
import GafferScene
import IECore
import imath

Gaffer.Metadata.registerValue( parent, "serialiser:milestoneVersion", 0, persistent=False )
Gaffer.Metadata.registerValue( parent, "serialiser:majorVersion", 57, persistent=False )
Gaffer.Metadata.registerValue( parent, "serialiser:minorVersion", 5, persistent=False )
Gaffer.Metadata.registerValue( parent, "serialiser:patchVersion", 0, persistent=False )

__children = {}

__children["Sphere"] = GafferScene.Sphere( "Sphere" )
parent.addChild( __children["Sphere"] )
__children["Sphere"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Group"] = GafferScene.Group( "Group" )
parent.addChild( __children["Group"] )
__children["Group"]["in"].addChild( GafferScene.ScenePlug( "in1", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Group"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["ContextVariables"] = Gaffer.ContextVariables( "ContextVariables" )
parent.addChild( __children["ContextVariables"] )
__children["ContextVariables"].setup( GafferScene.ScenePlug( "in", ) )
__children["ContextVariables"]["variables"].addChild( Gaffer.NameValuePlug( "", Gaffer.StringPlug( "value", defaultValue = '', flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ), True, "member1", Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic ) )
__children["ContextVariables"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Expression1"] = Gaffer.Expression( "Expression1" )
parent.addChild( __children["Expression1"] )
__children["Expression1"]["__out"].addChild( Gaffer.StringPlug( "p0", direction = Gaffer.Plug.Direction.Out, defaultValue = '', flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Expression1"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Group1"] = GafferScene.Group( "Group1" )
parent.addChild( __children["Group1"] )
__children["Group1"]["in"].addChild( GafferScene.ScenePlug( "in1", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Group1"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Group2"] = GafferScene.Group( "Group2" )
parent.addChild( __children["Group2"] )
__children["Group2"]["in"].addChild( GafferScene.ScenePlug( "in1", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Group2"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Transform"] = GafferScene.Transform( "Transform" )
parent.addChild( __children["Transform"] )
__children["Transform"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Expression"] = Gaffer.Expression( "Expression" )
parent.addChild( __children["Expression"] )
__children["Expression"]["__out"].addChild( Gaffer.FloatPlug( "p0", direction = Gaffer.Plug.Direction.Out, defaultValue = 0.0, flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Expression"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["PathFilter1"] = GafferScene.PathFilter( "PathFilter1" )
parent.addChild( __children["PathFilter1"] )
__children["PathFilter1"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Sphere"]["__uiPosition"].setValue( imath.V2f( 64.5644836, 49.8795815 ) )
__children["Group"]["in"][0].setInput( __children["ContextVariables"]["out"] )
__children["Group"]["__uiPosition"].setValue( imath.V2f( 64.5644836, 23.1075974 ) )
__children["ContextVariables"]["variables"]["member1"]["name"].setValue( 'blah' )
__children["ContextVariables"]["variables"]["member1"]["value"].setInput( __children["Expression1"]["__out"]["p0"] )
__children["ContextVariables"]["__uiPosition"].setValue( imath.V2f( 64.5644836, 31.2716599 ) )
__children["ContextVariables"]["in"].setInput( __children["Transform"]["out"] )
Gaffer.Metadata.registerValue( __children["ContextVariables"]["in"], 'noduleLayout:section', 'top' )
Gaffer.Metadata.registerValue( __children["ContextVariables"]["out"], 'noduleLayout:section', 'bottom' )
__children["Expression1"]["__uiPosition"].setValue( imath.V2f( 52.1233368, 30.7001762 ) )
__children["Group1"]["in"][0].setInput( __children["Group"]["out"] )
__children["Group1"]["__uiPosition"].setValue( imath.V2f( 64.5635834, 17.3728962 ) )
__children["Group2"]["in"][0].setInput( __children["Group1"]["out"] )
__children["Group2"]["__uiPosition"].setValue( imath.V2f( 64.5635834, 11.1523256 ) )
__children["Transform"]["enabled"].setValue( False )
__children["Transform"]["in"].setInput( __children["Sphere"]["out"] )
__children["Transform"]["filter"].setInput( __children["PathFilter1"]["out"] )
__children["Transform"]["transform"]["translate"]["x"].setInput( __children["Expression"]["__out"]["p0"] )
__children["Transform"]["__uiPosition"].setValue( imath.V2f( 64.7820435, 41.2328873 ) )
__children["Expression"]["__uiPosition"].setValue( imath.V2f( 54.2821312, 41.2320518 ) )
__children["PathFilter1"]["paths"].setValue( IECore.StringVectorData( [ '/sphere' ] ) )
__children["PathFilter1"]["__uiPosition"].setValue( imath.V2f( 77.7820435, 47.3149185 ) )
__children["Expression1"]["__engine"].setValue( 'python' )
__children["Expression1"]["__expression"].setValue( 'parent["__out"]["p0"] = \'\'\n\nfoo' )
__children["Expression"]["__engine"].setValue( 'python' )
__children["Expression"]["__expression"].setValue( 'parent["__out"]["p0"] = 0.0\nx' )

del __children
```

If you then run `root['Group2']['out'].transform( "/group/group/group/sphere" )`, it correctly reports an error with `Group2.out.transform`, because hash errors are reported on the most downstream plug evaluated, not the actual plug with the problem.  If you switch which of the two nodes with bogus expressions is enabled however, then the other one triggers an exception during compute, and that is reported with the correct plug.  I have not yet figured out the difference - I think I can see why the hash fails due to the exception being rethrowUnwrapped at each level, and recaught and reconverted into a new ProcessException - but I haven't figured out why the same thing doesn't go wrong during computes.

johnhaddon commented on src/Gaffer/ValuePlug.cpp:
```diff
@@ -284,10 +307,32 @@ class ValuePlug::HashProcess : public Process
+
+						if( result != check )
+						{
+							throw IECore::Exception( "Failed to clear hash cache.  Fix affects() code." );
```
I don't see how we can expect this Exception to get wrapped by a ProcessException for `plug`. Wrapping happens in `Process::handleException()`, but we're not inside a process for `plug` here - that process was created and destroyed inside `cache.get()`. That said, I haven't looked at your other example yet - going to take a look at that now.

johnhaddon replied:
Wait, scratch that! I've expanded the surrounding code and seen the manual					`ProcessException::wrapCurrentException( processKey.plug, Context::current(), staticType );`

johnhaddon commented on src/Gaffer/ValuePlug.cpp:
```diff
@@ -138,6 +138,8 @@ struct ThreadStateFixer
 
 #endif
 
+uint64_t DIRTY_COUNT_RANGE_HALF = ((~(uint64_t)0) >> 1 ) + 1;
```
const!

johnhaddon commented:
We also need to update the wraparound threshold here : https://github.com/GafferHQ/gaffer/pull/3836/commits/4a5add48a698460c0fb7aa43709f5193f4e38ba6#diff-e573c259f370615173e4c22604e60c1dL853

danieldresser-ie commented:
There were some weird details in getting this working - it definitely needs close review.  But at least it's ready for review now.

johnhaddon commented on src/Gaffer/ValuePlug.cpp:
```diff
@@ -391,13 +391,17 @@ class ValuePlug::HashProcess : public Process
 			{
-				g_legacyGlobalDirtyCount = 0;
-				clearCache();
+				uint64_t c = g_legacyGlobalDirtyCount;
+				while( !g_legacyGlobalDirtyCount.compare_exchange_weak( c, ( c + 1 ) % DIRTY_COUNT_RANGE_HALF ) )
+				{
+				}
+		
+				if( c == DIRTY_COUNT_RANGE_HALF - 1 )
+				{
+					clearCache();
+				}
```
I don't think this provides thread safety. Imagine two threads, each with their own node graph. One makes an edit and detects wraparound, then the other makes an edit and launches a compute _before_ the first thread has got to `clearCache()`.

Given that we don't really make use of different graphs in different threads (in the UI, all graphs are managed on the UI thread) it's debatable how much we should care about this. I suggest we either simplify things and make no pretencee of thread safety, or we use a mutex, knowing that we shouldn't hit it often because we shouldn't be in Legacy mode, and even then contention with another thread is unlikely.

danieldresser-ie replied:
Implemented using spin mutex as of a82db4b

johnhaddon commented on src/Gaffer/ValuePlug.cpp:
```diff
@@ -1142,21 +1215,16 @@ void ValuePlug::dirty()
-		// while the bugs are fixed.
-		g_clearHashCacheOnDirty
-	)
+	if( m_dirtyCount >= DIRTY_COUNT_RANGE_HALF )
```
I think I mentioned this before, but there is existing code in `~ValuePlug` that also needs to be updated to wrap at `DIRTY_COUNT_RANGE_HALF` rather than `numeric_limits<uint64_t::max()>`.

danieldresser-ie replied:
I've taken a swing at this in a82db4b, though it was a bit more complicated than I expected.

As far as I can tell, the previous version of this code would completely fail most of the time.  For simplicity, assume max_uint == 1000.  You have some plug with a dirty count of 998.  Then it gets dirtied 4 times, and then deleted.  It now has a m_dirtyCount of 2, and would not have triggered the previous special case branch.  The dirty propogation that pushes it to wrap would have done a clearCache anyway, so that might be OK, but the value of g_dirtyCountEpoch might still be set at 900 or something, due to the `std::max`.  This would appear to permanently disable g_dirtyCountEpoch ( with a chance to reenable if you happen to land exactly on max_uint the next time it wraps around ).

The version I've now written I think is correct at least, though it does seem like a substantial performance hazard.  I hadn't really thought through how bad the rollover scenario could get, particularly because it's plausible to have many plugs with similar dirty counts - it's not one global clearCache, it's one per plug ... and with this new code, likely 2 per plug.

I thought about fixing that extra clear per plug, and I think it would involve:
* only clear when you land exactly on the wraparound, like before
* in cases where you have wrapped, but not cleared, you need to do a std::max with the current g_dirtyCountEpoch, but ignoring values that haven't wrapped yet ( just ignoring values > DIRTY_COUNT_RANGE_HALF / 2 would do the trick )

That would increase the complexity of this code though, so I've currently just done the simplest thing that I think should give accurate results.

johnhaddon commented on src/GafferTest/MultiplyNode.cpp:
```diff
@@ -45,13 +45,15 @@ GAFFER_GRAPHCOMPONENT_DEFINE_TYPE( MultiplyNode )
 	addChild( new IntPlug( "op2" ) );
 	addChild( new IntPlug( "product", Plug::Out ) );
+
+	m_brokenAffects = brokenAffects;
```
Better to use an initialiser than an assignment in the constructor.

danieldresser-ie replied:
Addressed in dd25539

johnhaddon commented on src/Gaffer/ValuePlug.cpp:
```diff
@@ -445,68 +438,51 @@ class ValuePlug::HashProcess : public Process
 
 		static IECore::MurmurHash globalCacheGetter( const HashProcessKey &key, size_t &cost )
 		{
-			try
```
This is a nice simplification to the code - thanks for suggesting it.

danieldresser-ie commented on include/Gaffer/Private/IECorePreview/LRUCache.h:
```diff
@@ -105,9 +105,7 @@ class LRUCache : private boost::noncopyable
-		LRUCache( GetterFunction getter );
-		LRUCache( GetterFunction getter, Cost maxCost );
-		LRUCache( GetterFunction getter, RemovalCallback removalCallback, Cost maxCost );
+		LRUCache( GetterFunction getter, Cost maxCost = 500, RemovalCallback removalCallback = nullRemovalCallback );
```
Addressed in ff178a4

danieldresser-ie commented on include/Gaffer/Private/IECorePreview/LRUCache.h:
```diff
@@ -146,6 +146,8 @@ class LRUCache : private boost::noncopyable
 		/// Returns the current cost of all cached items.
 		Cost currentCost() const;
 
+		static void nullRemovalCallback( const Key &key, const Value &value );
```
Addressed in ff178a4.  Note that this required adding an explicit check for empty callbacks, and that I've now removed nullRemovalCallback since it was private and unused.

danieldresser-ie commented on src/Gaffer/ValuePlug.cpp:
```diff
@@ -223,6 +229,25 @@ bool spawnsTasks( const HashProcessKey &key )
+		}
+	}
+	return ValuePlug::HashCacheMode::Standard;
+}
```
Addressed in a82db4b

johnhaddon replied:
We also need to update the messages we output to use the same case - Legacy rather than LEGACY etc.

danieldresser-ie commented on src/Gaffer/ValuePlug.cpp:
```diff
@@ -260,9 +285,11 @@ class ValuePlug::HashProcess : public Process
-			const HashProcessKey processKey( p, plug, Context::current(), p->m_dirtyCount, computeNode, computeNode ? computeNode->hashCachePolicy( p ) : CachePolicy::Uncached );
 
-			if( processKey.cachePolicy == CachePolicy::Uncached )
+			ValuePlug::CachePolicy cachePolicy = computeNode ? computeNode->hashCachePolicy( p ) : CachePolicy::Uncached;
```
Addressed in a82db4b

danieldresser-ie commented on src/Gaffer/ValuePlug.cpp:
```diff
@@ -284,10 +311,43 @@ class ValuePlug::HashProcess : public Process
 
 				// And then look up the result in our cache.
-
+				
```
I think this should be sorted out now

danieldresser-ie commented on src/Gaffer/ValuePlug.cpp:
```diff
@@ -284,10 +311,43 @@ class ValuePlug::HashProcess : public Process
+					{
+						HashProcessKey legacyProcessKey( processKey );
+						legacyProcessKey.dirtyCount = g_legacyGlobalDirtyCount + DIRTY_COUNT_RANGE_HALF;
+						IECore::MurmurHash check = threadData.cache.get( legacyProcessKey );
```
Addressed in a82db4b

danieldresser-ie commented on src/Gaffer/ValuePlug.cpp:
```diff
@@ -284,10 +311,43 @@ class ValuePlug::HashProcess : public Process
+							// approach, which can be done by throwing and then immediately wrapping
+							try
+							{
+								throw IECore::Exception( "Failed to clear hash cache.  Fix affects() code." );
```
Addressed in a82db4b

danieldresser-ie commented on src/Gaffer/ValuePlug.cpp:
```diff
@@ -329,6 +389,29 @@ class ValuePlug::HashProcess : public Process
+			clearCache();
+		}
+
+        static ValuePlug::HashCacheMode getHashCacheMode()
```
Addressed in a82db4b

danieldresser-ie commented on src/Gaffer/ValuePlug.cpp:
```diff
@@ -1142,21 +1215,16 @@ void ValuePlug::dirty()
-		// while the bugs are fixed.
-		g_clearHashCacheOnDirty
-	)
+	if( m_dirtyCount >= DIRTY_COUNT_RANGE_HALF )
```
Addressed in a82db4b

danieldresser-ie commented:
Hopefully this all looks reasonable ( unless I've misunderstood what's happening in ~ValuePlug ).

I've done fixup commits, but I've also thrown out some commits ( managing dropping commits that contained alternate approaches in the middle of this code as fixups to revert changes felt like a merging nightmare ), hopefully the history mismatch isn't a problem.

danieldresser-ie commented:
Oh, I see I need to fix a Python 3 test thing.  But if I do that in the fixup commit it would throw off the hash I mention in my comment ... I guess I could do multiple fixups?  I'll worry about it tomorrow.

danieldresser-ie commented:
OK, I've fixed the test for Python 3, addressed the "LEGACY" vs "Legacy" message, rebased on your split out PR, and squashed this down.

Should be about good to go.  Regarding your parting comment this morning, is the basic idea that plugs actually getting destructed is rare enough that we may not be worried too much about the peformance in the weird wrap-around special cases?  That's kinda fair, and makes me think that the way I've currently got it might be reasonable ( the previous code which I'm pretty sure could have put g_dirtyCountEpoch in an invalid state would still be dangerous ).

danieldresser-ie commented:
Put up two more commits addressing the two things John pointed out this morning:  fixed a copy-paste error, and got rid of the mutex on g_legacyGlobalDirtyCount. It's now back to using an atomic, and I think using the compare_exchange_weak gives us adequate thread safety, since it seems like it should guarantee, even under race conditions:
1) Every call to dirtyLegacyCache does result in a new value being assigned to the count.
2) The count is always in the range of 0 <= count < RANGE_HALF
3) At least one clearCache() will be issued shortly BEFORE the count wraps around.

Contention could result in duplicate clearCache() calls, but that shouldn't be a big problem.

johnhaddon commented on src/Gaffer/ValuePlug.cpp:
```diff
@@ -798,10 +890,17 @@ ValuePlug::~ValuePlug()
+		nextDirtyCount = 0;
+	}
+
+	if( nextDirtyCount > g_dirtyCountEpoch )
```
Doesn't this new version lead to a lot of unnecessary clearing of the cache?

- Make two plugs
- Dirty one N times and then delete it. `g_dirtyCountEpoch == N`
- Delete the second plug. `nextDirtyCount <= g_dirtyCountEpoch` and we clear the cache unnecessarily

I think this scenario is actually very likely. I think the commonest cause of plug deletions is the use of temporary utility plugs like the ones that the UIEditor makes to edit presets, and those will naturally have very low dirty counts.

Am I missing something?

danieldresser-ie replied:
Dangit, you're totally right.

I started thinking about how we can detect a wraparound just based on the current values of g_dirtyCountEpoch and m_dirtyCount, but there isn't really enough information.

New proposal for different approach:

ValuePlug::dirty() should reset g_dirtyCountEpoch to 0 when it does a clearCache() - if the cache is being cleared anyway, there's no reason not to reset it.

Instead of replicating any of the logic in dirty, ~ValuePlug should just start by calling dirty().  This will increment m_dirtyCount, taking care of wrapping and potentially clearing.

Then ~ValuePlug can just do:
`while( !g_dirtyCountEpoch.compare_exchange_weak( epoch, std::max( epoch, m_dirtyCount ) ) )`

This is safe because if a wraparound happened, g_dirtyCountEpoch would have been reset by dirty(), so we don't need to worry about the `max` causing epoch to stay stuck high.

danieldresser-ie replied:
No, wait, this still doesn't really work great ...

The biggest problem is that it feels like both of us have been thinking about this like the dirty counts on all plugs advance together, so that we catch the point in time where a wrap happens, and then everything gets back to normal.

But in reality, if a plug reaches DIRTY_COUNT_RANGE_HALF, a far more likely scenario is that the dirty counts of the other plugs in the scene are evenly distributed between 0 and DIRTY_COUNT_RANGE_HALF ... meaning cache clears are likely to continue occuring continually for the rest of the session - nothing is ever going to force the dirty counts back into the same portion of the range, once they've spread out.

The ping ponging between updates of plugs that have wrapped and haven't wrapped I think also might cause problems for the approach I just described that would cause excessive cacheClears ... though maybe this isn't that big a problem in itself, now that I've realized you would start getting constant cacheClears anyways without even deleting any plugs?

I'm not sure how we actually solve this without setting all m_dirtyCount on all plugs to 0 whenever we do a cacheClear.

danieldresser-ie commented:
I've squashed everything, and reverted ~ValuePlug to more like what it was before, with a comment that this is WIP.  ( I didn't bother keeping this as a separate commit, since you can just take a look at ~ValuePlug ).  Should be good to merge while we figure out what to do with m_dirtyCount.

johnhaddon commented:
Thanks Daniel - I've pushed one more revision that fixes a couple of minor typos, but also adds a `dirtyLegacyCache()` call to `~ValuePlug`, as without that I think we have a bug - correct? Will merge once CI is done.
